src/coreclr/nativeaot/Runtime/unix/PalCreateDump.cpp

```diff
@@ -0,0 +1,583 @@
+#include <pthread_np.h>
+#define PlatformGetCurrentThreadId() (uint32_t)pthread_getthreadid_np()
+#elif defined(__NetBSD__)
+#include <lwp.h>
+#define PlatformGetCurrentThreadId() (uint32_t)_lwp_self()
+#else
+#define PlatformGetCurrentThreadId() (uint32_t)pthread_self()
+#endif
+
+const size_t MaxUnsigned32BitDecString = STRING_LENGTH("4294967295");
+const size_t MaxUnsigned64BitDecString = STRING_LENGTH("18446744073709551615");
+
+/*++
+Function:
+    FormatInt
+
+    Helper function to format an uint32 as a string.
+
+--*/
+static
+char*
+FormatInt(uint32_t value)
+{
+    char* buffer = (char*)malloc(MaxUnsigned32BitDecString + 1);
+    if (buffer != nullptr)
+    {
+        if (snprintf(buffer, MaxUnsigned32BitDecString, "%" PRIu32, value) < 0)
+        {
+            free(buffer);
+            buffer = nullptr;
+        }
+    }
+    return buffer;
+}
+
+/*++
+Function:
+    FormatInt64
+
+    Helper function to format an uint64 as a string.
+
+--*/
+static
+char*
+FormatInt64(uint64_t value)
+{
+    char* buffer = (char*)malloc(MaxUnsigned64BitDecString + 1);
+    if (buffer != nullptr)
+    {
+        if (snprintf(buffer, MaxUnsigned64BitDecString, "%" PRIu64, value) < 0)
+        {
+            free(buffer);
+            buffer = nullptr;
+        }
+    }
+    return buffer;
+}
+
+static const int UndefinedDumpType = 0;
+
+/*++
+Function
+  BuildCreateDumpCommandLine
+
+Abstract
+  Builds the createdump command line from the arguments.
+
+Return
+  true - succeeds, false - fails
+
+--*/
+static
+bool
+BuildCreateDumpCommandLine(
+    const char** argv,
+    const char* dumpName,
+    const char* logFileName,
+    int dumpType,
+    uint32_t flags)
+{
+    if (g_szCreateDumpPath == nullptr || g_ppidarg == nullptr)
+    {
+        return false;
+    }
+
+    int argc = 0;
+    argv[argc++] = g_szCreateDumpPath;
+
+    if (dumpName != nullptr)
+    {
+        argv[argc++] = "--name";
+        argv[argc++] = dumpName;
+    }
+
+    switch (dumpType)
+    {
+        case 1: argv[argc++] = "--normal";
+            break;
+        case 2: argv[argc++] = "--withheap";
+            break;
+        case 3: argv[argc++] = "--triage";
+            break;
+        case 4: argv[argc++] = "--full";
+            break;
+        default:
+            break;
+    }
+
+    if (flags & GenerateDumpFlagsLoggingEnabled)
+    {
+        argv[argc++] = "--diag";
+    }
+
+    if (flags & GenerateDumpFlagsVerboseLoggingEnabled)
+    {
+        argv[argc++] = "--verbose";
+    }
+
+    if (flags & GenerateDumpFlagsCrashReportEnabled)
+    {
+        argv[argc++] = "--crashreport";
+    }
+
+    if (flags & GenerateDumpFlagsCrashReportOnlyEnabled)
+    {
+        argv[argc++] = "--crashreportonly";
+    }
+
+    if (logFileName != nullptr)
+    {
+        argv[argc++] = "--logtofile";
+        argv[argc++] = logFileName;
+    }
+
+    argv[argc++] = "--nativeaot";
+    argv[argc++] = g_ppidarg;
+    argv[argc++] = nullptr;
+    assert(argc < MAX_ARGV_ENTRIES);
```

Comment:
This should not be a in-debug runtime check. You'll only hit it if you test all variants.